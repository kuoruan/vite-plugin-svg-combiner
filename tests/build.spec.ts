import { type RollupOutput } from "rollup";
import { build, mergeConfig } from "vite";

import svgCombiner from "@/index";

import viteConfig from "./vite.config";

describe("build", () => {
  it("basic", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/basic.ts",
          },
        },
        plugins: [svgCombiner()],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });

  it("sub directory", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/sub-directory.ts",
          },
        },
        plugins: [svgCombiner()],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });

  it("minify", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/minify.ts",
          },
        },
        plugins: [svgCombiner()],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });

  it("glob import", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/glob.ts",
          },
        },
        plugins: [svgCombiner()],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });
});

describe("build file", () => {
  it("basic", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/basic.ts",
          },
        },
        plugins: [
          svgCombiner({
            emitFile: true,
          }),
        ],
      }),
    )) as RollupOutput;

    expect(data.output[1]).toMatchSnapshot();
  });

  it("custom file name", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/minify.ts",
          },
        },
        plugins: [
          svgCombiner({
            emitFile: "custom.svg",
          }),
        ],
      }),
    )) as RollupOutput;

    expect(data.output[1]).toMatchSnapshot();
  });
});

describe("custom filter", () => {
  it("include files", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/include.ts",
          },
        },
        plugins: [svgCombiner({ include: /window/ })],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });

  it("exclude files", async () => {
    const data = (await build(
      mergeConfig(viteConfig, {
        build: {
          rollupOptions: {
            input: "/fixtures/exclude.ts",
          },
        },
        plugins: [svgCombiner({ exclude: /window/ })],
      }),
    )) as RollupOutput;

    expect(data.output[0].code).toMatchSnapshot();
  });
});
