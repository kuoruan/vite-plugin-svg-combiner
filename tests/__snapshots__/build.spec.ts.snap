// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build > basic 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="add" viewBox="0 -960 960 960"><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80z"/></symbol>', "add");
"
`;

exports[`build > custom element id 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function setElementId(id) {
  elementId = id;
  if (root) {
    root.id = id;
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="add" viewBox="0 -960 960 960"><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80z"/></symbol>', "add");
setElementId("custom-id");
"
`;

exports[`build > glob import 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="add" viewBox="0 -960 960 960"><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80z"/></symbol>', "add");
addSymbol('<symbol id="remove" viewBox="0 -960 960 960"><path d="M200-440v-80h560v80z"/></symbol>', "remove");
addSymbol('<symbol id="verdaccio-small" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100"><defs><path id="a" d="M48 17.6 32.8 48H24L.4.8h15.2l12.8 25.6 4.4-8.8z"/><path id="d" d="M50.8 12H35.6L41.2.8h15.2z"/><path id="f" d="M32.8 48H24L.4.8h15.2l20.377 40.89z"/><filter id="c" width="139.9%" height="140.3%" x="-20%" y="-11.7%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><filter id="e" width="191.3%" height="269.6%" x="-45.7%" y="-49.1%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter></defs><g fill="none" fill-rule="evenodd"><rect width="100" height="100" fill="#F7F8F6" rx="37"/><g transform="translate(22 29)"><use xlink:href="#a" fill="#000" filter="url(#c)"/><use xlink:href="#a" fill="#405236"/><path stroke="#405236" stroke-width="2.4" d="M46.058 18.8H33.542L28.4 29.083 14.858 2H2.342l22.4 44.8h7.316z"/></g><g transform="translate(22 29)"><use xlink:href="#d" fill="#000" filter="url(#e)"/><use xlink:href="#d" fill="#CD4000"/><path stroke="#CD4000" stroke-width="2.4" d="m50.058 10.8 4.4-8.8H41.942l-4.4 8.8z"/></g><g transform="translate(22 29)"><use xlink:href="#f" fill="#4A5E3F"/><path stroke="#405236" stroke-width="2.4" d="m32.06 46.8 2.575-5.112L14.857 2H2.342l22.4 44.8z"/></g><path stroke="#CD4000" stroke-linecap="square" stroke-width="2.4" d="M59.6 31h15.221M55.6 35h15.221M51.6 39.8h15.221"/></g></symbol>', "verdaccio-small");
"
`;

exports[`build > minify 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="verdaccio-small" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100"><defs><path id="a" d="M48 17.6 32.8 48H24L.4.8h15.2l12.8 25.6 4.4-8.8z"/><path id="d" d="M50.8 12H35.6L41.2.8h15.2z"/><path id="f" d="M32.8 48H24L.4.8h15.2l20.377 40.89z"/><filter id="c" width="139.9%" height="140.3%" x="-20%" y="-11.7%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><filter id="e" width="191.3%" height="269.6%" x="-45.7%" y="-49.1%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter></defs><g fill="none" fill-rule="evenodd"><rect width="100" height="100" fill="#F7F8F6" rx="37"/><g transform="translate(22 29)"><use xlink:href="#a" fill="#000" filter="url(#c)"/><use xlink:href="#a" fill="#405236"/><path stroke="#405236" stroke-width="2.4" d="M46.058 18.8H33.542L28.4 29.083 14.858 2H2.342l22.4 44.8h7.316z"/></g><g transform="translate(22 29)"><use xlink:href="#d" fill="#000" filter="url(#e)"/><use xlink:href="#d" fill="#CD4000"/><path stroke="#CD4000" stroke-width="2.4" d="m50.058 10.8 4.4-8.8H41.942l-4.4 8.8z"/></g><g transform="translate(22 29)"><use xlink:href="#f" fill="#4A5E3F"/><path stroke="#405236" stroke-width="2.4" d="m32.06 46.8 2.575-5.112L14.857 2H2.342l22.4 44.8z"/></g><path stroke="#CD4000" stroke-linecap="square" stroke-width="2.4" d="M59.6 31h15.221M55.6 35h15.221M51.6 39.8h15.221"/></g></symbol>', "verdaccio-small");
"
`;

exports[`build > sub directory 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="remove" viewBox="0 -960 960 960"><path d="M200-440v-80h560v80z"/></symbol>', "remove");
addSymbol('<symbol id="close" viewBox="0 -960 960 960"><path d="m256-200-56-56 224-224-224-224 56-56 224 224 224-224 56 56-224 224 224 224-56 56-224-224z"/></symbol>', "close");
"
`;

exports[`build file > basic 1`] = `
{
  "fileName": "svg-sprite.svg",
  "name": undefined,
  "needsCodeReference": false,
  "source": "<svg xmlns="http://www.w3.org/2000/svg"><symbol id="add" viewBox="0 -960 960 960"><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80z"/></symbol></svg>",
  "type": "asset",
}
`;

exports[`build file > custom file name 1`] = `
{
  "fileName": "custom.svg",
  "name": undefined,
  "needsCodeReference": false,
  "source": "<svg xmlns="http://www.w3.org/2000/svg"><symbol id="verdaccio-small" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100"><defs><path id="a" d="M48 17.6 32.8 48H24L.4.8h15.2l12.8 25.6 4.4-8.8z"/><path id="d" d="M50.8 12H35.6L41.2.8h15.2z"/><path id="f" d="M32.8 48H24L.4.8h15.2l20.377 40.89z"/><filter id="c" width="139.9%" height="140.3%" x="-20%" y="-11.7%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><filter id="e" width="191.3%" height="269.6%" x="-45.7%" y="-49.1%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter></defs><g fill="none" fill-rule="evenodd"><rect width="100" height="100" fill="#F7F8F6" rx="37"/><g transform="translate(22 29)"><use xlink:href="#a" fill="#000" filter="url(#c)"/><use xlink:href="#a" fill="#405236"/><path stroke="#405236" stroke-width="2.4" d="M46.058 18.8H33.542L28.4 29.083 14.858 2H2.342l22.4 44.8h7.316z"/></g><g transform="translate(22 29)"><use xlink:href="#d" fill="#000" filter="url(#e)"/><use xlink:href="#d" fill="#CD4000"/><path stroke="#CD4000" stroke-width="2.4" d="m50.058 10.8 4.4-8.8H41.942l-4.4 8.8z"/></g><g transform="translate(22 29)"><use xlink:href="#f" fill="#4A5E3F"/><path stroke="#405236" stroke-width="2.4" d="m32.06 46.8 2.575-5.112L14.857 2H2.342l22.4 44.8z"/></g><path stroke="#CD4000" stroke-linecap="square" stroke-width="2.4" d="M59.6 31h15.221M55.6 35h15.221M51.6 39.8h15.221"/></g></symbol></svg>",
  "type": "asset",
}
`;

exports[`custom filter > exclude files 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="add" viewBox="0 -960 960 960"><path d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80z"/></symbol>', "add");
addSymbol('<symbol id="remove" viewBox="0 -960 960 960"><path d="M200-440v-80h560v80z"/></symbol>', "remove");
addSymbol('<symbol id="verdaccio-small" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100"><defs><path id="a" d="M48 17.6 32.8 48H24L.4.8h15.2l12.8 25.6 4.4-8.8z"/><path id="d" d="M50.8 12H35.6L41.2.8h15.2z"/><path id="f" d="M32.8 48H24L.4.8h15.2l20.377 40.89z"/><filter id="c" width="139.9%" height="140.3%" x="-20%" y="-11.7%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><filter id="e" width="191.3%" height="269.6%" x="-45.7%" y="-49.1%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter></defs><g fill="none" fill-rule="evenodd"><rect width="100" height="100" fill="#F7F8F6" rx="37"/><g transform="translate(22 29)"><use xlink:href="#a" fill="#000" filter="url(#c)"/><use xlink:href="#a" fill="#405236"/><path stroke="#405236" stroke-width="2.4" d="M46.058 18.8H33.542L28.4 29.083 14.858 2H2.342l22.4 44.8h7.316z"/></g><g transform="translate(22 29)"><use xlink:href="#d" fill="#000" filter="url(#e)"/><use xlink:href="#d" fill="#CD4000"/><path stroke="#CD4000" stroke-width="2.4" d="m50.058 10.8 4.4-8.8H41.942l-4.4 8.8z"/></g><g transform="translate(22 29)"><use xlink:href="#f" fill="#4A5E3F"/><path stroke="#405236" stroke-width="2.4" d="m32.06 46.8 2.575-5.112L14.857 2H2.342l22.4 44.8z"/></g><path stroke="#CD4000" stroke-linecap="square" stroke-width="2.4" d="M59.6 31h15.221M55.6 35h15.221M51.6 39.8h15.221"/></g></symbol>', "verdaccio-small");
"
`;

exports[`custom filter > include files 1`] = `
"let root;
let elementId = "svg-sprite";
if (typeof document !== "undefined") {
  root = document.createElementNS("http://www.w3.org/2000/svg", "svg");
  root.id = elementId;
  root.style.position = "absolute";
  root.style.left = "-1px";
  root.style.top = "-1px";
  root.style.width = "0px";
  root.style.height = "0px";
  root.style.visibility = "hidden";
  const insertSvgSprite = () => {
    document.body.insertBefore(root, document.body.firstChild);
  };
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", insertSvgSprite);
  } else {
    insertSvgSprite();
  }
}
function addSymbol(symbol, symbolId) {
  if (!root)
    return;
  const symbolNode = root.querySelector(\`#\${symbolId}\`);
  if (symbolNode) {
    try {
      symbolNode.remove();
    } catch {
      root.removeChild(symbolNode);
    }
  }
  root.insertAdjacentHTML("beforeend", symbol);
}
addSymbol('<symbol id="close" viewBox="0 -960 960 960"><path d="m256-200-56-56 224-224-224-224 56-56 224 224 224-224 56 56-224 224 224 224-56 56-224-224z"/></symbol>', "close");
"
`;
